server/src/main/java/org/elasticsearch/inference/ChunkedInferenceServiceResults.java

```diff
@@ -8,6 +8,20 @@
 
 package org.elasticsearch.inference;
 
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.xcontent.XContent;
+
+import java.util.AbstractMap;
+import java.util.Iterator;
+
 public interface ChunkedInferenceServiceResults extends InferenceServiceResults {
 
+    Iterator<Chunk> chunksAsMatchedTextAndByteReference(XContent xcontent);
```

Comment:
This looks much better as `SemanticText` doesn't have to cast to each separate interface 👍 

I wonder - would it be possible to have a common Chunks superclass / interface for:
- ChunkedTextExpansionResults.ChunkedResult
- EmbeddingChunk

That way, the chunks could be written using a common method in the superclass instead of implementing a new record and method.

Do you think that's doable?

Reply:
It seems possible to me for that superclass/interface to exist, although I am not really sure what the point of `ChunkedNlpInferenceResults` is, so I could be wrong. 

I'm not sure what you mean about implementing a new record and method? Each class needs to implement chunksAsMatchedTextAndByteReference because the logic to do so depends on attributes of the subclass.

Reply:
> Each class needs to implement chunksAsMatchedTextAndByteReference because the logic to do so depends on attributes of the subclass

Yes, my point was that the current chunk classes (ChunkedTextExpansionResults.ChunkedResult
EmbeddingChunk) could implement that behaviour instead of needing an additional Chunk record for doing this.

I think we can deal with this as a separate issue for further refactoring, I'll try to come up with an alternative. The main point was to let `SemanticText` deal transparently with new classes, and this PR achieves that 👍